test/widgets/message_list_test.dart

```diff
@@ -371,11 +377,32 @@ void main() {
       //   https://github.com/dart-lang/sdk/issues/28985 (about DateTime.now, not timezone)
       //   https://github.com/dart-lang/sdk/issues/44928 (about the Dart implementation's own internal tests)
       // For this test, just accept outputs corresponding to any possible timezone.
-      tester.widget(find.textContaining(RegExp("2022-12-1[89]")));
-      tester.widget(find.textContaining(RegExp("2023-08-2[23]")));
+      tester.widget(find.textContaining(RegExp("Dec 1[89](, 2022)?")));
+      tester.widget(find.textContaining(RegExp("Aug 2[23](, 2023)?")));
     });
   });
 
+  group('formatHeaderDate', () {
+    final zulipLocalizations = GlobalLocalizations.zulipLocalizations;
+    final now = DateTime(2023, 1, 10, 12);
+    final testCases = [
+      [DateTime(2023, 1, 10, 12),     zulipLocalizations.temporalNounToday],
+      [DateTime(2023, 1, 10, 0),      zulipLocalizations.temporalNounToday],
+      [DateTime(2023, 1, 10, 23, 59), zulipLocalizations.temporalNounToday],
+      [DateTime(2023, 1, 9, 23, 59),  zulipLocalizations.temporalNounYesterday],
+      [DateTime(2023, 1, 9, 0),       zulipLocalizations.temporalNounYesterday],
+      [DateTime(2023, 1, 8),          "Jan 8"],
+      [DateTime(2022, 12, 31),        "Dec 31, 2022"],
+      // Times in the future should show up as a normal date.
+      [DateTime(2023, 1, 11, 12),     "Jan 11"],
```

Comment:
> … Oh I see. This actually matches the logic we _currently_ have in web. What happened is that we accidentally altered that behavior in a commit that was described as just replacing a dependency, [zulip/zulip@9896782](https://github.com/zulip/zulip/commit/9896782fd14de1e3b573cb232762988af0e1fe6b) . This condition had matched the comment, but no longer does:
> 
> ```diff
> -    } else if (is_older_year) {
> +    } else if (time.getFullYear() !== today.getFullYear()) {
>          // For long running servers, searching backlog can get ambiguous
>          // without a year stamp. Only show year if message is from an older year
> ```

FYI @andersk as the author of that change. Though the root issue here is that we apparently never had a test for that case.

Reply:
I dug into this a bit this morning. It looks like `format_time_modern` actually hasn't been used since it was reverted back in March 2022 in 7bc0e70693f49ea8faef2578158362491fe5470c . See https://chat.zulip.org/#narrow/stream/137-feedback/topic/recent.20topics.20timestamps/near/1337670 for background.

Web is currently using `relative_time_string_from_date` for this label (the date for last message time in recent topics).

```js
export function relative_time_string_from_date({
    date,
    current_date = new Date(),
}: {
    date: Date;
    current_date?: Date; 
}): string {
    const minutes = differenceInMinutes(current_date, date);
    if (minutes <= 2) {
        return $t({defaultMessage: "Just now"});
    }
```
and sadly it looks like this function will always return `Just now` for any times from the future. I can also confirm by changing my computer clock to the previous day and every `time` in my inbox shows `Just now`.

I think showing times in the future is weird. Even for `format_time_modern` it shows the date label if the future time even a minute ahead, despite being on the same day (so it could be Dec 1 and the label will say `Dec 1, 2023` where I think it should continue to say `Today`). I think it's quite likely for devices to drift a few minutes and hit this condition.

Open to further discussion, but for now I'm making a change so that for future dates, dates that are today still show up as `Today` but dates further off will have the year attached.

Reply:
> t looks like `format_time_modern` actually hasn't been used since it was reverted back in March 2022 in [zulip/zulip@7bc0e70](https://github.com/zulip/zulip/commit/7bc0e70693f49ea8faef2578158362491fe5470c) .

Hmm, yeah, I remember that conversation now. Ah well.


> and sadly it looks like this function will always return `Just now` for any times from the future.

Wow, yeah, that seems wrong.


> for future dates, dates that are today still show up as `Today` but dates further off will have the year attached.

Sure. Given that we're only showing to calendar-day granularity anyway, that's the behavior I'd expect here.